### ISSUE TITLE
Orientation bias initial population

### ISSUE THREAD

**Issue by L30nardoSV** [ccsb-scripps/AutoDock-GPU#36]
From @diogo.martins:

Using the x-ray ligand as input (as opposed to using a randomized conformation/orientation) results in better energies and a greater probability of finding the best solution. Therefore, docking performance is unpredictable, because favorable inputs result in better performance, while worst inputs result in worse performance. Docking performance should be independent of the conformation and orientation of the input ligand. Indeed, an AutoDock user complained about this problem in a former version of AutoDock, which was fixed afterwards. 

It is very likely (but not certain) that fixing the orientation bias in the initial population will solve this problem.

According to https://arxiv.org/pdf/physics/0506177.pdf:
> If the orientation is given in axis-angle space, s, then the axis, s/ |s|, should be chosen uniformly on S 2 , and the rotation angle, |s|, should be sampled from [0, π] with probability (2/π) sin2 (|s| /2).

The first part - _choosing uniformly on S 2_ - is equivalent to choosing a random point on a sphere surface, see http://mathworld.wolfram.com/SpherePointPicking.html (note that _theta_ and _phi_ have inverted meaning with respect to ocladock notation).

The second part - sampling _angle_ with probability (2/π) sin2 (|angle| /2) - is trickier. A look-up table might be the easiest way to implement it. As far as I know, there's no analytical function that provides such probability distribution.

**Comment by L30nardoSV** [comment]
From @diogo.martins:

Change of plans: instead of implementing the algorithm described above, the following algorithm is implemented in 0c18c0a7232cb00644f1b62a366179209e7bb1cb.

1. Generate a random uniform quaternion using the method from [Shoemake](http://planning.cs.uiuc.edu/node198.html)
2. Convert the quaternion to ocladock's axis-angle representation (`phi`, `theta`, `rotangle`). 
3. Remove the randomization of reference orientation [host/src/getparameters.cpp, l. 753-755](https://git.esa.informatik.tu-darmstadt.de/docking/ocladock/blob/96a0f0f595ec1c61c6f2f4759543a358f4657d26/host/src/getparameters.cpp#L753). As far as I understand, this is not necessary.

##### Validation 1 of 2

Run ocladock with `-nrun 1000` and `-nev 1`, to generate an initial population for visualization. The image below shows the distribution of one ligand atom before (left) and after (right) this implementation. On the left panel, the ligand atom can be found more often in the top right corner.

![validation1](/uploads/7c37186cd4238f66d0fe8064002a2260/validation1.png)


##### Validation 2 of 2 FAILING

Docking performance for system "1ig3". In the plots below, the x-axis is the number of energy evals `-nev`. The y-axis is the probability of success, defined as the fraction of docking runs for which the best solution is within 1 kcal/mol from the best solution ever observed. The plots with one black line (top row) use the x-ray ligand as input, while the plots with 10 lines use randomized ligands as input (bottom row). After the above algorithm is implemented (right side), the performance does depend even more on the input, which is the opposite of the desired behavior.

![AMPLIFIED-PROBLEM](/uploads/1deb7fd063fba1b9b58a7c8f92492f53/AMPLIFIED-PROBLEM.png)

**Comment by L30nardoSV** [comment]
From @diogo.martins:

Problem fixed in commit e10762fdeac25611a81df36f5b3db19a544ef6b6, see figure below. Performance is virtually independent of input orientation.

This was fixed by (re-) introducing randomization of the reference orientation [host/src/getparameters.cpp l. 801-829](https://github.com/ccsb-scripps/AutoDock-GPU/blob/e10762fdeac25611a81df36f5b3db19a544ef6b6/host/src/getparameters.cpp#L801). I don't understand why this is necessary.

@solis: I mixed spaces and tabs, and repeated code, sorry about that.

![SOLVED-ISSUE36](https://git.esa.informatik.tu-darmstadt.de/docking/ocladock/uploads/35aa0f00c509ef9e9d748595411c0bc6/SOLVED-ISSUE36.png)
